### DISCUSSION TITLE
Make autofocus an user accessibility preference

### DISCUSSION THREAD

**Discussion by mmikkel** [craftcms/cms#12921] [Ideas] [answered]
Craft 4.4 removed `autofocus` from the CP (https://github.com/craftcms/cms/pull/12332, https://github.com/craftcms/cms/pull/12406). This makes sense, as the feature can represent an accessibility issue, or cause otherwise unwanted behaviour (https://github.com/craftcms/cms/discussions/12324).

On the other hand; for authors who haven't experienced any issues with the autofocusing behaviour previously and who might prefer it, this change arguably makes the Craft AX a bit more cumbersome and will likely be seen as a regression.

Would it make sense to bring `autofocus` back, with a new accessibility preference to let users opt in or out?

**Comment by khalwat** [comment]
Agree, at least make it an option, with the current behavior (prior to the removal of autofocus) being the default imo.

**Reply by thupsi** [reply]
Is there a reason autofocus should be the default when not creating new elements?

**Reply by mmikkel** [reply]
FWIW, assuming this theoretical autofocus accessibility preference was also added to the [accessibilityDefaults](https://craftcms.com/docs/4.x/config/general.html#accessibilitydefaults) setting, developers would be free to set whatever default they deem most appropriate, for each project.

**Reply by tomkiss** [reply]
Adding it to the accessibility config seems like a good idea.
Although, as a feature generally regarded as problematic for accessibility and often discouraged, I think not having it on by default would be the most appropriate thing to do.
I also think if there were a config setting, having two settings; one for creation and one for editing would be helpful.

**Reply by khalwat** [reply]
Having it default to what it has been for the past 10 years or so of Craft's life, and what content authors have gotten used to, is probably best IMO.

Then allowing people to opt-in to accessibility enhancements if they need them, and you have a win-win.

**Reply by tomkiss** [reply]
The issue with autofocus is that its presence may pose an accessibility _problem_ for some users (and for Craft in meeting WCAG success criterion). 
So "opting-in" to _disable_ it would imply that the user gets the least accessibility compliant version by default.

**Reply by brandonkelly** [reply]
We are going to be introducing a new onboarding wizard for new CP users, which will give them the ability to set their accessibility preferences right up front, before they jump into the rest of the CP. So it doesn’t really matter if the default preference is less accessible, as it will be very short-lived if undesired.

**Reply by mmikkel** [reply]
Looking forward to that wizard (IMO it should probably pop up for existing users as well, the first time they log in after the site has been upgraded to a version that includes it).

Also, assuming the wizard will have preferences' defaults checked (which is nice for people who may not care too much, and just zip through it), I hope to see the `accessibilityDefaults` setting retained. Even if the wizard lets users immediately set their individual preferences, it's still going to be useful for the developer to be able to define what the defaults should be, per project.

**Reply by khalwat** [reply]
Great plan -- similar to what Apple does!

**Reply by tomkiss** [reply]
Sounds great Brandon 💯

**Comment by brandonkelly** [comment, marked as answer]
I agree that this was a little heavy-handed, and should have come with a new “Reduce autofocus” accessibility preference. Reverted the changes for now (81a217e06bf964113a2349b8433295dce787110c), and we will add a preference for it in a future release.

> I also think if there were a config setting, having two settings; one for creation and one for editing would be helpful.

Two separate preferences feels a little overkill to me… even if autofocusing is less helpful when editing existing elements, it’s not going to hurt you if the title field were autofocused.

**Reply by mmikkel** [reply]
That's great, thanks a lot Brandon 🎉

**Reply by brandonkelly** [reply]
Craft 4.4.4 is out now, bringing back autofocus.

**Comment by brandonkelly** [comment]
Craft 5.0.0-alpha.10 was just released with a new “Disable autofocus” user accessibility preference (#14251).
